File: expansion-packs/bmad-research-dev/templates/servers/zotero/get-item.ts
/**
 * Zotero Get Item Server Wrapper
 * Wraps mcp__zotero__get_item_metadata and related MCP tools
 *
 * Usage:
 *   import { getItem } from './servers/zotero/get-item.ts'
 *   const item = await getItem("ABC123XYZ")
 */

interface ItemMetadata {
  key: string;
  version: number;
  itemType: string;
  title: string;
  creators: Array<{
    creatorType: string;
    firstName?: string;
    lastName: string;
    name?: string;
  }>;
  abstractNote?: string;
  publicationTitle?: string;
  date?: string;
  DOI?: string;
  url?: string;
  pages?: string;
  volume?: string;
  issue?: string;
  tags: Array<{ tag: string; type?: number }>;
  collections: string[];
  relations: Record<string, any>;
  dateAdded: string;
  dateModified: string;
  extra?: string;
}

interface Note {
  key: string;
  itemType: 'note';
  note: string; // HTML content
  tags: Array<{ tag: string }>;
  dateAdded: string;
  dateModified: string;
}

interface Annotation {
  key: string;
  itemType: 'annotation';
  annotationType: 'highlight' | 'note' | 'image';
  annotationText?: string;
  annotationComment?: string;
  annotationColor?: string;
  annotationPageLabel?: string;
  annotationPosition?: string;
  dateAdded: string;
  dateModified: string;
}

interface Attachment {
  key: string;
  itemType: 'attachment';
  title: string;
  filename?: string;
  path?: string;
  contentType?: string;
  url?: string;
  dateAdded: string;
  dateModified: string;
}

/**
 * Get full metadata for a Zotero item by key
 *
 * @param itemKey - Zotero item key
 * @param format - Output format (json, bibtex)
 * @returns Item metadata
 *
 * @example
 * // Get item metadata
 * const item = await getItem("ABC123XYZ")
 * console.log(item.title)
 *
 * @example
 * // Get BibTeX citation
 * const bibtex = await getItem("ABC123XYZ", "bibtex")
 * console.log(bibtex)
 */
export async function getItem(
  itemKey: string,
  format: 'json' | 'bibtex' = 'json'
): Promise<ItemMetadata | string> {
  console.log(`[Zotero] Getting item: ${itemKey} (format: ${format})`)

  // Call the actual mcp__zotero__get_item_metadata tool
  const result = await globalThis.mcp__zotero__get_item_metadata({
    item_key: itemKey,
    format
  });

  console.log(`[Zotero] Retrieved item: ${typeof result === 'string' ? 'BibTeX' : result.title}`)

  return result;
}

/**
 * Get multiple items in parallel
 */
export async function getItems(
  itemKeys: string[],
  format: 'json' | 'bibtex' = 'json'
): Promise<Array<ItemMetadata | string>> {
  console.log(`[Zotero] Getting ${itemKeys.length} items in parallel...`)

  const items = await Promise.all(
    itemKeys.map(key => getItem(key, format))
  );

  console.log(`[Zotero] Retrieved ${items.length} items`)

  return items;
}

/**
 * Get item's full text content
 */
export async function getItemFulltext(itemKey: string): Promise<string> {
  console.log(`[Zotero] Getting full text for item: ${itemKey}`)

  const result = await globalThis.mcp__zotero__get_item_fulltext({
    item_key: itemKey
  });

  console.log(`[Zotero] Retrieved full text (${result.length} characters)`)

  return result;
}

/**
 * Get item's child items (attachments, notes)
 */
export async function getItemChildren(itemKey: string): Promise<{
  attachments: Attachment[];
  notes: Note[];
  annotations: Annotation[];
}> {
  console.log(`[Zotero] Getting children for item: ${itemKey}`)

  const children = await globalThis.mcp__zotero__get_item_children({
    item_key: itemKey
  });

  // Categorize children by type
  const attachments: Attachment[] = [];
  const notes: Note[] = [];
  const annotations: Annotation[] = [];

  children.forEach((child: any) => {
    if (child.itemType === 'attachment') {
      attachments.push(child);
    } else if (child.itemType === 'note') {
      notes.push(child);
    } else if (child.itemType === 'annotation') {
      annotations.push(child);
    }
  });

  console.log(`[Zotero] Found ${attachments.length} attachments, ${notes.length} notes, ${annotations.length} annotations`)

  return { attachments, notes, annotations };
}

/**
 * Get annotations for an item
 */
export async function getAnnotations(itemKey: string): Promise<Annotation[]> {
  console.log(`[Zotero] Getting annotations for item: ${itemKey}`)

  const result = await globalThis.mcp__zotero__get_annotations({
    item_key: itemKey
  });

  console.log(`[Zotero] Found ${result.length} annotations`)

  return result;
}

/**
 * Get notes for an item
 */
export async function getNotes(itemKey: string): Promise<Note[]> {
  console.log(`[Zotero] Getting notes for item: ${itemKey}`)

  const result = await globalThis.mcp__zotero__get_notes({
    item_key: itemKey
  });

  console.log(`[Zotero] Found ${result.length} notes`)

  return result;
}

/**
 * Search within notes and annotations
 */
export async function searchNotes(
  query: string,
  searchAnnotations: boolean = true
): Promise<Array<{ itemKey: string; content: string; type: 'note' | 'annotation' }>> {
  console.log(`[Zotero] Searching notes for: "${query}"`)

  const result = await globalThis.mcp__zotero__search_notes({
    query,
    search_annotations: searchAnnotations
  });

  console.log(`[Zotero] Found ${result.length} matching notes/annotations`)

  return result;
}

/**
 * Create a new note for an item
 */
export async function createNote(
  itemKey: string,
  noteContent: string,
  tags?: string[]
): Promise<Note> {
  console.log(`[Zotero] Creating note for item: ${itemKey}`)

  const result = await globalThis.mcp__zotero__create_note({
    item_key: itemKey,
    note: noteContent,
    tags: tags || []
  });

  console.log(`[Zotero] Created note: ${result.key}`)

  return result;
}

/**
 * Get comprehensive item details (metadata + children + fulltext)
 */
export async function getItemComplete(itemKey: string): Promise<{
  metadata: ItemMetadata;
  fulltext?: string;
  attachments: Attachment[];
  notes: Note[];
  annotations: Annotation[];
}> {
  console.log(`[Zotero] Getting complete details for item: ${itemKey}`)

  // Fetch everything in parallel
  const [metadata, children, fulltext] = await Promise.all([
    getItem(itemKey),
    getItemChildren(itemKey),
    getItemFulltext(itemKey).catch(() => undefined) // Fulltext may not be available
  ]);

  console.log(`[Zotero] Retrieved complete item details`)

  return {
    metadata: metadata as ItemMetadata,
    fulltext,
    ...children
  };
}

/**
 * Export items as BibTeX
 */
export async function exportBibTeX(itemKeys: string[]): Promise<string> {
  console.log(`[Zotero] Exporting ${itemKeys.length} items as BibTeX`)

  const bibtexEntries = await Promise.all(
    itemKeys.map(key => getItem(key, 'bibtex'))
  );

  const bibtex = bibtexEntries.join('\n\n');

  console.log(`[Zotero] Exported ${itemKeys.length} BibTeX entries`)

  return bibtex;
}

/**
 * Analyze item's research value
 */
export async function analyzeItemResearchValue(
  itemKey: string
): Promise<{
  hasFulltext: boolean;
  hasAnnotations: boolean;
  hasNotes: boolean;
  annotationCount: number;
  noteCount: number;
  hasCode: boolean;
  hasData: boolean;
  readiness: 'ready' | 'partial' | 'metadata-only';
}> {
  const [fulltext, children] = await Promise.all([
    getItemFulltext(itemKey).catch(() => undefined),
    getItemChildren(itemKey)
  ]);

  const hasFulltext = !!fulltext;
  const hasAnnotations = children.annotations.length > 0;
  const hasNotes = children.notes.length > 0;

  // Check for code/data availability
  const fulltextLower = fulltext?.toLowerCase() || '';
  const hasCode = /github|code.*available|repository/i.test(fulltextLower);
  const hasData = /dataset.*available|data.*provided/i.test(fulltextLower);

  // Determine readiness for research use
  let readiness: 'ready' | 'partial' | 'metadata-only' = 'metadata-only';
  if (hasFulltext && (hasAnnotations || hasNotes)) {
    readiness = 'ready';
  } else if (hasFulltext) {
    readiness = 'partial';
  }

  return {
    hasFulltext,
    hasAnnotations,
    hasNotes,
    annotationCount: children.annotations.length,
    noteCount: children.notes.length,
    hasCode,
    hasData,
    readiness
  };
}

/**
 * Get items with your annotations/notes (well-read papers)
 */
export async function getAnnotatedItems(limit?: number): Promise<Array<{
  item: ItemMetadata;
  annotationCount: number;
  noteCount: number;
}>> {
  // This requires searching all items and checking for annotations
  // In practice, you'd use zotero_search_items with specific criteria
  console.log(`[Zotero] Finding annotated items...`)

  // This is a placeholder - actual implementation would query Zotero
  // for items with annotations/notes
  throw new Error('getAnnotatedItems requires advanced search - use search with filters instead');
}

File: expansion-packs/bmad-research-dev/templates/servers/zotero/get_item.py
"""
Zotero Get Item Server - Direct Web API
Makes direct HTTPS requests to api.zotero.org (NO MCP)

API Docs: https://www.zotero.org/support/dev/web_api/v3/basics

Usage:
    from servers.zotero.get_item import get_item
    item = get_item("ABC123XYZ")
"""

import os
import requests
from typing import List, Dict, Optional, Any, Literal, Union
from dotenv import load_dotenv

# Load environment variables
load_dotenv()


def get_config() -> Dict[str, str]:
    """Get environment configuration"""
    api_key = os.getenv('ZOTERO_API_KEY', '')
    user_id = os.getenv('ZOTERO_USER_ID', '')

    if not api_key or not user_id:
        raise ValueError(
            'Zotero credentials not found. Set ZOTERO_API_KEY and ZOTERO_USER_ID in .env file.'
        )

    return {'apiKey': api_key, 'userId': user_id}


def api_request(
    endpoint: str,
    params: Optional[Dict[str, str]] = None,
    format: Literal['json', 'bibtex'] = 'json'
) -> Union[Dict[str, Any], List[Dict[str, Any]], str]:
    """Make direct API request to Zotero"""
    config = get_config()

    if params is None:
        params = {}

    # Build URL
    url = f"https://api.zotero.org/users/{config['userId']}{endpoint}"

    # Add format parameter if specified
    if format != 'json':
        params['format'] = format

    print(f"[Zotero API] GET {endpoint}")

    # Make HTTPS request
    headers = {
        'Zotero-API-Key': config['apiKey'],
        'Zotero-API-Version': '3'
    }

    response = requests.get(url, headers=headers, params=params)

    if not response.ok:
        raise Exception(f"Zotero API error: {response.status_code} {response.reason}")

    # Return as text for bibtex, JSON for others
    if format == 'bibtex':
        return response.text
    else:
        return response.json()


def get_item(
    item_key: str,
    format: Literal['json', 'bibtex'] = 'json'
) -> Union[Dict[str, Any], str]:
    """
    Get full metadata for a Zotero item by key

    Args:
        item_key: Zotero item key
        format: Output format (json, bibtex)

    Returns:
        Item metadata

    Examples:
        # Get item metadata
        item = get_item("ABC123XYZ")
        print(item['data']['title'])

        # Get BibTeX citation
        bibtex = get_item("ABC123XYZ", "bibtex")
        print(bibtex)
    """
    print(f"[Zotero] Getting item: {item_key} (format: {format})")

    result = api_request(f'/items/{item_key}', format=format)

    if isinstance(result, str):
        print("[Zotero] Retrieved BibTeX citation")
    else:
        print(f"[Zotero] Retrieved item: {result.get('data', {}).get('title', '')}")

    return result


def get_items(
    item_keys: List[str],
    format: Literal['json', 'bibtex'] = 'json'
) -> List[Union[Dict[str, Any], str]]:
    """Get multiple items in parallel"""
    print(f"[Zotero] Getting {len(item_keys)} items in parallel...")

    items = [get_item(key, format) for key in item_keys]

    print(f"[Zotero] Retrieved {len(items)} items")

    return items


def get_item_fulltext(item_key: str) -> str:
    """Get item's full text content"""
    print(f"[Zotero] Getting full text for item: {item_key}")

    try:
        result = api_request(f'/items/{item_key}/fulltext')

        if isinstance(result, dict):
            content = result.get('content', '')
            print(f"[Zotero] Retrieved full text ({len(content)} characters)")
            return content
        else:
            print(f"[Zotero] No fulltext available for {item_key}")
            raise Exception(f"No fulltext available for {item_key}")
    except Exception as error:
        # Fulltext may not be available
        print(f"[Zotero] No fulltext available for {item_key}")
        raise error


def get_item_children(item_key: str) -> Dict[str, List[Dict[str, Any]]]:
    """Get item's child items (attachments, notes, annotations)"""
    print(f"[Zotero] Getting children for item: {item_key}")

    children = api_request(f'/items/{item_key}/children')

    # Categorize children by type
    attachments = []
    notes = []
    annotations = []

    if isinstance(children, list):
        for child in children:
            item_type = child.get('data', {}).get('itemType')
            if item_type == 'attachment':
                attachments.append(child)
            elif item_type == 'note':
                notes.append(child)
            elif item_type == 'annotation':
                annotations.append(child)

    print(f"[Zotero] Found {len(attachments)} attachments, {len(notes)} notes, {len(annotations)} annotations")

    return {
        'attachments': attachments,
        'notes': notes,
        'annotations': annotations
    }


def get_annotations(item_key: str) -> List[Dict[str, Any]]:
    """Get annotations for an item"""
    print(f"[Zotero] Getting annotations for item: {item_key}")

    children = get_item_children(item_key)
    annotations = children['annotations']

    print(f"[Zotero] Found {len(annotations)} annotations")

    return annotations


def get_notes(item_key: str) -> List[Dict[str, Any]]:
    """Get notes for an item"""
    print(f"[Zotero] Getting notes for item: {item_key}")

    children = get_item_children(item_key)
    notes = children['notes']

    print(f"[Zotero] Found {len(notes)} notes")

    return notes


def get_item_complete(item_key: str) -> Dict[str, Any]:
    """Get comprehensive item details (metadata + children + fulltext)"""
    print(f"[Zotero] Getting complete details for item: {item_key}")

    # Fetch metadata
    metadata = get_item(item_key)

    # Fetch children
    children = get_item_children(item_key)

    # Fetch fulltext (may not be available)
    try:
        fulltext = get_item_fulltext(item_key)
    except:
        fulltext = None

    print("[Zotero] Retrieved complete item details")

    return {
        'metadata': metadata,
        'fulltext': fulltext,
        **children
    }


def export_bibtex(item_keys: List[str]) -> str:
    """Export items as BibTeX"""
    print(f"[Zotero] Exporting {len(item_keys)} items as BibTeX")

    bibtex_entries = [get_item(key, 'bibtex') for key in item_keys]

    bibtex = '\n\n'.join(str(entry) for entry in bibtex_entries)

    print(f"[Zotero] Exported {len(item_keys)} BibTeX entries")

    return bibtex


def analyze_item_research_value(item_key: str) -> Dict[str, Any]:
    """Analyze item's research value"""
    import re

    # Get fulltext
    try:
        fulltext = get_item_fulltext(item_key)
        has_fulltext = True
    except:
        fulltext = None
        has_fulltext = False

    # Get children
    children = get_item_children(item_key)

    has_annotations = len(children['annotations']) > 0
    has_notes = len(children['notes']) > 0

    # Check for code/data availability
    fulltext_lower = (fulltext or '').lower()
    has_code = bool(re.search(r'github|code.*available|repository', fulltext_lower))
    has_data = bool(re.search(r'dataset.*available|data.*provided', fulltext_lower))

    # Determine readiness for research use
    if has_fulltext and (has_annotations or has_notes):
        readiness = 'ready'
    elif has_fulltext:
        readiness = 'partial'
    else:
        readiness = 'metadata-only'

    return {
        'hasFulltext': has_fulltext,
        'hasAnnotations': has_annotations,
        'hasNotes': has_notes,
        'annotationCount': len(children['annotations']),
        'noteCount': len(children['notes']),
        'hasCode': has_code,
        'hasData': has_data,
        'readiness': readiness
    }
